Build MCP server  + connect client LLM guides in Next.js (#2312)
clerk/clerk-docs: docs/references/express/build-mcp-service.mdx

---
title: Build an MCP service in your Express application with Clerk
description: Learn how to build an MCP service using Clerk's OAuth server in your Express application.
---

If you're building an application and want to allow large language model (LLM) apps to securely access user data within your app, you'll need to set up a Model Context Protocol (MCP) service. MCP enables external tools like LLM-based apps to request user permission to access specific data from your service. Refer to the [dedicated MCP section](/docs/oauth/how-clerk-implements-oauth#building-a-model-context-protocol-mcp-service-using-clerk-s-o-auth-server) for more information.

There are two parties involved in MCP:

- The `client` is the LLM application that wants to access another service on a user's behalf. For example, Claude would be the "client" if it wants to get access to Gmail.

> - The `server` is the system that hosts the protected resources the client wants to access. In this example, this would be Gmail. This is sometimes referred to as the "resource server" or "MCP server".

Reviewer: May be worth adding some sort of note about how "client" and "server" are normally used to refer to frontend and backend, and they are specifically different in this context, which can be confusing but is important to remember

  <CodeBlockTabs options={["npm", "yarn", "pnpm", "bun"]}>
    ```bash {{ filename: 'terminal' }}
    npm install @modelcontextprotocol/sdk @clerk/mcp-tools
    ```

    ```bash {{ filename: 'terminal' }}
    yarn add @modelcontextprotocol/sdk @clerk/mcp-tools
    ```

    ```bash {{ filename: 'terminal' }}
    pnpm add @modelcontextprotocol/sdk @clerk/mcp-tools
    ```

    ```bash {{ filename: 'terminal' }}
    bun add @modelcontextprotocol/sdk @clerk/mcp-tools
    ```
  </CodeBlockTabs>

  ## Set up your Express app with Clerk and MCP imports

  The following code shows the required imports and setup needed to implement an MCP server with Clerk.

  ```ts
  import 'dotenv/config'
  import { type MachineAuthObject, clerkClient, clerkMiddleware } from '@clerk/express'
  import {
    mcpAuthClerk,
    protectedResourceHandlerClerk,
    streamableHttpHandler,
  } from '@clerk/mcp-tools/express'
  import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js'
  import express from 'express'

  const app = express()
  app.use(clerkMiddleware())
  app.use(express.json())

  app.listen(3000)

>   ```

Reviewer: Missing semicolons here

Author: The formatting script doesn't accept semicolons - seems they're not included in all code snippets in our docs

  ## Create your MCP server and define tools

>   To let external LLM-powered tools securely interact with your app, you need to define an MCP server and expose at least one tool. A tool is a capability your MCP server offers. The following example registers a single tool called `get_clerk_user_data` that retrieves information about the authenticated Clerk user.

Reviewer: ```suggestion
  To let external LLM-powered tools securely interact with your app, you need to define an MCP server, and expose one or more [resources](https://modelcontextprotocol.io/docs/concepts/resources), [prompts](https://modelcontextprotocol.io/docs/concepts/prompts), and/or [tools](https://modelcontextprotocol.io/docs/concepts/tools). The following example registers a single tool called `get_clerk_user_data` that retrieves information about the authenticated Clerk user.
```

  1. The `McpServer` function is used to create a new MCP server with a name and version.
  1. The `server.tool()` function is used to define a tool that external LLM-based apps can invoke. Each tool includes:

     - A name (`get_clerk_user_data`).
     - A description of what the tool does.
     - Input parameters (none in this case).

>      - A callback function that extracts the user's auth info from the OAuth token, fetches the user data using `clerk.users.getUser()` and returns it wrapped in a `content` array, which is MCP's expected response format.

Reviewer: ```suggestion
     - A function that represents the implementation of the too, in this case, it extracts the user's auth info from the OAuth token, fetches the user's data using `clerk.users.getUser()` and returns it in MCP's expected response format.
```

  ```ts
  const server = new McpServer({
    name: 'test-server',
    version: '0.0.1',
  })

  server.tool(
    'get_clerk_user_data',
    'Gets data about the Clerk user that authorized this request',
    {},
    async (_, { authInfo }) => {
      const clerkAuthInfo = authInfo as unknown as MachineAuthObject<'oauth_token'>
      if (!clerkAuthInfo?.userId) {
        console.error(authInfo)
        return {
          content: [{ type: 'text', text: 'Error: user not authenticated' }],
        }
      }
      const user = await clerkClient.users.getUser(clerkAuthInfo.userId)
      return {
        content: [{ type: 'text', text: JSON.stringify(user) }],
      }
    },
  )
  ```

  ## Secure and expose your MCP server

>   Now that your MCP server and tools are defined, the next step is to secure your endpoints and expose them according to the MCP specification. To comply with the MCP specification, your server must expose OAuth resource metadata at a `well-known` endpoint. This metadata allows clients (like LLM tools) to know the authentication requirements and token formats your server accepts.

Reviewer: ```suggestion
  Now that your MCP server and tools are defined, the next step is to secure your endpoints and expose them according to the MCP specification. To comply with the MCP specification, your server must expose [OAuth protected resource metadata](https://datatracker.ietf.org/doc/html/rfc9728#section-4.1) at a specific endpoint (`.well-known/oauth-protected-resource`) This metadata allows clients discover where to authenticate and some details about how the authentication service works.
```

  Clerk provides prebuilt helpers that handle this for you:

>   - `mcpAuthClerk`: Authentication middleware that automatically verifies the `Authorization` header using Clerk-issued OAuth tokens.

Reviewer: ```suggestion
  - `mcpAuthClerk`: Authentication middleware that automatically verifies the `Authorization` header using Clerk-issued OAuth tokens. If unauthorized, it returns a www-authenticate header [in accordance with the mcp specification](https://modelcontextprotocol.io/specification/2025-06-18/basic/authorization#authorization-server-location).
```

>   - `protectedResourceHandlerClerk`: Express handler that serves the OAuth protected resource metadata, required for dynamic client registration.

Reviewer: ```suggestion
  - `protectedResourceHandlerClerk`: Express handler that serves OAuth protected resource metadata.
```

>   - `streamableHttpHandler`: Express handler that connects your MCP server to incoming requests using the HTTP streaming transport.

Reviewer: ```suggestion
  - `streamableHttpHandler`: Express handler that connects your MCP server to incoming requests using the [streamable HTTP transport](https://modelcontextprotocol.io/docs/concepts/transports#streamable-http).
```

  > [!NOTE]
  > For a more in-depth explanation of these helpers, see the [MCP Express reference](https://github.com/clerk/mcp-tools/tree/main/express).

  To secure your endpoints and expose your MCP server to a client:

  ```ts
  app.post('/mcp', mcpAuthClerk, streamableHttpHandler(server))
  app.get('/.well-known/oauth-protected-resource', protectedResourceHandlerClerk)
  ```

>   ## Connecting your client to the MCP server

Reviewer: I think this is good for now, but it's likely we'll want to pull the "connecting LLMs to Clerk MCP servers" docs out into their own page that we can link from each one of these guides, since it will be the same for all of them

Author: Created a partial for this for now
